File: java-kata/src/main/java/org/acmebank/katas/bankaccount/StatementFilter.java
package org.acmebank.katas.bankaccount;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class StatementFilter {
    
    public static List<Transaction> depositsOnly(List<Transaction> transactions) {
        return transactions.stream()
            .filter(Transaction::isDeposit)
            .collect(Collectors.toList());
    }
    
    public static List<Transaction> withdrawalsOnly(List<Transaction> transactions) {
        return transactions.stream()
            .filter(Transaction::isWithdrawal)
            .collect(Collectors.toList());
    }
    
    public static List<Transaction> byDate(List<Transaction> transactions, LocalDate date) {
        return transactions.stream()
            .filter(transaction -> transaction.getDate().toLocalDate().equals(date))
            .collect(Collectors.toList());
    }
    
    public static List<Transaction> byDateRange(List<Transaction> transactions, LocalDate startDate, LocalDate endDate) {
        return transactions.stream()
            .filter(transaction -> {
                LocalDate transactionDate = transaction.getDate().toLocalDate();
                return !transactionDate.isBefore(startDate) && !transactionDate.isAfter(endDate);
            })
            .collect(Collectors.toList());
    }
    
    public static List<Transaction> applyFilter(List<Transaction> transactions, Predicate<Transaction> filter) {
        return transactions.stream()
            .filter(filter)
            .collect(Collectors.toList());
    }
}

File: scala-kata/src/main/scala/org/acmebank/katas/bankaccount/StatementFilter.scala
package org.acmebank.katas.bankaccount

import java.time.LocalDate

object StatementFilter {
  
  def depositsOnly(transactions: List[Transaction]): List[Transaction] = 
    transactions.filter(_.isDeposit)
  
  def withdrawalsOnly(transactions: List[Transaction]): List[Transaction] = 
    transactions.filter(_.isWithdrawal)
  
  def byDate(transactions: List[Transaction], date: LocalDate): List[Transaction] = 
    transactions.filter(_.date.toLocalDate == date)
  
  def byDateRange(transactions: List[Transaction], startDate: LocalDate, endDate: LocalDate): List[Transaction] = 
    transactions.filter { transaction =>
      val transactionDate = transaction.date.toLocalDate
      !transactionDate.isBefore(startDate) && !transactionDate.isAfter(endDate)
    }
  
  def applyFilter(transactions: List[Transaction], filter: Transaction => Boolean): List[Transaction] = 
    transactions.filter(filter)
}